# Pandas. Домашнее задание

В домашнем задании есть обязательные и дополнительные задачи.

*Обязательные задачи* нужно сделать, чтобы проверить, что вы действительно усвоили материал модуля. Они обязательны к загрузке на проверку преподавателю.

*Дополнительные задачи* для тех, кто чувствует, что нужно ещё немного потренироваться и порешать задачки по работе с Pandas. Вы также можете загрузить их на проверку преподавателю и получить по ним обратную связь.

Удачи!



---



---



**Цели домашнего задания:**
1. Научиться работать с описательными статистиками набора данных.
2. Научиться выбирать данные по условиям.
3. Научиться объединять несколько датафреймов в один.


**Что входит в домашнее задание:**
1. *Обязательные задачи*:

17 задач с использованием:
- набора рекламных данных, показывающих, щелкнул ли конкретный пользователь интернета рекламу на веб-сайте компании или нет,
- дополнительного набора данных, описывающего пользователей, посещающих веб-сайт компании.
2. *Дополнительные задачи*:

3 задачи для отработки исследования данных.


**Критерии оценки:**
- программа выдаёт верный ответ на заданном наборе данных;
- код читабелен: переменным даны осмысленные названия, отступы и правила расстановки пробелов соблюдены.

**Как отправить задание на проверку**

Выполните задачи, сохраните изменения при помощи опции Save and Checkpoint из вкладки меню File или кнопки Save and Checkpoint на панели инструментов. Отправьте через форму на платформе итоговый файл Jupyter Notebook (в формате .ipynb) или ссылку на него.



---


---



**Обязательные задачи**

В течение всего домашнего задания вы будете работать с несколькими наборами данных:
1. Набор рекламных данных, показывающих, щёлкнул ли конкретный пользователь рекламу на сайте компании или нет. Состоит из трёх разных файлов: `advertising_1.csv`, `advertising_2.csv`, `advertising_3.csv`.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Daily Time Spent on Site**: Среднее ежедневное время нахождения пользователя на сайте в минутах.
- **Daily Internet Usage**: Среднее ежедневное время нахождения пользователя в интернете в минутах.
- **Ad Topic Line**: Заголовок рекламного объявления.
- **Clicked on Ad**: Кликнул пользователь на рекламу или нет.

Дополнительный к рекламным данным датасет, который описывает пользователей, посещающих сайт компании. Состоит из одного файла: users.csv.

Описание столбцов датасета:
- **Number**: Номер пользователя в системе.
- **Age**: Возраст пользователя.
- **Male**: Является ли пользователь мужчиной.
- **City**: Город пользователя.
- **Country**: Страна пользователя.
- **Area Income**: Средний доход в месте жительства пользователя.


***Задача 1***

Загрузите таблицу из файла `advertising_1.csv` и сохраните её в датафрейм `adv1_df`. Укажите в качестве индекса столбец `Number`. Выведите на экран первые десять строк.


In [2]:
import pandas as pd

In [6]:
# Ваш код здесь
adv1_df = pd.read_csv('advertising_1.csv', index_col='Number')

Daily Time Spent on Site  Daily Internet Usage  \
Number                                                   
5                          68.95                256.09   
6                          80.23                193.77   
8                          74.15                245.89   
9                            NaN                225.58   
10                         59.99                226.74   
...                          ...                   ...   
525                        46.61                136.18   
526                        63.26                120.46   
527                        79.16                202.90   
528                        67.94                128.16   
529                        79.91                230.18   

                                Ad Topic Line  Clicked on Ad  
Number                                                        
5          Cloned 5thgeneration orchestration              0  
6          Monitored national standardization              0  
8       Triple-buffered reciprocal time-frame              0  
9               Robust logistical utilization              0  
10            Sharable client-driven software              0  
...                                       ...            ...  
525       Progressive intermediate throughput              1  
526             Customizable holistic archive              1  
527           Compatible intermediate concept              0  
528      Assimilated next generation firmware              1  
529        Total zero administration software              0  

[500 rows x 4 columns]

***Задача 2***

Выведите размер датафрейма `adv1_df`.

In [7]:
adv1_df.shape

(500, 4)

***Задача 3***

Выведите cреднее ежедневное время нахождения в интернете пользователя под номером 8.

***Задача 4***

Загрузите таблицу из файла `advertising_2.csv` и сохраните её в датафрейм `adv2_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран строки для пользователей с номерами с 533-го по 536-й.

In [ ]:
# Ваш код здесь

***Задача 5***

Выведите сводную таблицу описательных статистик для датафрейма `adv2_df`, вызвав необходимый метод на всём датафрейме. 

Глядя на вывод, напишите, столбец с каким названием не содержит незаполненных значений.

*Подсказка:*  В сводной таблице не содержится информации о количестве строк всего датафрейма, поэтому для сначала распечатайте общее количество строк датафрейма.

*Рекомендация:* Кроме изученного метода `describe` для описания датафрейма и получения информации о пропущенных значениях полезен метод [`info`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.info.html).

In [ ]:
# Ваш код здесь

In [ ]:
# Ваш код здесь
print('Не содержит незаполненных значений столбец:', ваше_значение)

***Задача 6***

Добавьте к датафрейму `adv1_df` в конец датафрейм `adv2_df` с помощью метода `append`. 

Сохраните объединенный датафрейм в новой переменной `adv12_df` и распечатайте его.

In [ ]:
# Ваш код здесь

***Задача 7***

Для объединённого датафрейма `adv12_df` распечатайте среднее время, проводимое пользователем на сайте компании.

In [ ]:
# Ваш код здесь

***Задача 8***

Из объединённого датафрейма `adv12_df` выберите и распечатайте строки только с теми пользователями, для которых неизвестно среднее ежедневное время, проводимое пользователем на сайте компании.

In [ ]:
# Ваш код здесь

***Задача 9***

Загрузите таблицу из файла `advertising_3.csv` и сохраните её в датафрейм `adv3_df`. Укажите в качестве индекса стоблец `Number`. Выведите на экран столбцы `Ad Topic Line` и `Clicked on Ad`.

In [ ]:
# Ваш код здесь

***Задача 10***

Добавьте к датафрейму `adv12_df` в конец датафрейм `adv3_df` с помощью метода `concat`. Явно пропишите значение параметра `axis`.

Сохраните объединённый датафрейм в новой переменной `adv123_df` и распечатайте его.

In [ ]:
# Ваш код здесь

***Задача 11***

Для объединённого датафрейма `adv123_df` выберите только те строки, в которых пользователи кликнули на рекламное сообщение. Сохраните полученный результат в датафрейм `success_adv_df`, распечатайте размер нового датафрейма.

In [ ]:
# Ваш код здесь

***Задача 12***

Загрузите таблицу с данными пользователей из файла `users.csv` и сохраните её в датафрейм `users_df`. Индекс не указывайте.

Распечатайте сам датафрейм и сводную таблицу со статистиками для него.

In [ ]:
# Ваш код здесь

***Задача 13***

Глядя на статистики `users_df`, укажите значения, удовлетворяющие следующим условиям:
1. Возраст самого молодого пользователя сайта: *ваш_ответ*
2. Максимальный средний доход в месте жительства пользователя сайта: *ваш_ответ*
3. Средний возраст пользователя сайта: *ваш_ответ*
4. Количество заполненных значений для колонки, содержащей информацию о том, является пользователь мужчиной или нет: *ваш_ответ*
5. Минимальный номер пользователя: *ваш_ответ*

В этом задании не требуется выбирать по условиям и распечатывать значения, достаточно внимательно посмотреть на статистики и скопировать нужные значения согласно вопросам.

***Задача 14***

К полученному датафрейму с успешными рекламными сообщениями `success_adv_df` добавьте справа информацию из датафрейма с пользователями `users_df` и сохраните в датафрейм `success_full_df`. В полученном датафрейме должны остаться только те пользователи, номера которых есть в обоих датафреймах. Для этого воспользуйтесь методом `merge` и пропишите явно верные параметры.

Затем распечатайте датафрейм.

*Важно:*

Объединять датафреймы с помощью `merge` можно и по индексам, и по столбцам с разными названиями. 
Так как общие значения содержатся в столбце `Number`, но в датафрейме `users_df` данный столбец является обычным столбцом, а в `success_adv_df` является индексом, перед тем как объединять, переведите столбец `Number` в `success_adv_df` в обычный столбец.

Для этого перед объединением пропишите:
```
success_adv_df = success_adv_df.reset_index()
```

[Подробнее о reset_index](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.reset_index.html)

In [ ]:
# Ваш код здесь

***Задача 15***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, отобразите сводную таблицу со статистиками для строковых колонок (`Ad Topic Line`, `City`, `Country`).


In [ ]:
# Ваш код здесь

***Задача 16***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите страну и количество вхождений этой страны в датафрейм.


In [ ]:
# Ваш код здесь

***Задача 17***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных.


In [ ]:
# Ваш код здесь

***Задача 18***

Для объединённого датафрейма с пользователями, которые кликнули на рекламное сообщение `success_full_df`, выведите строки с данными пользователей из страны, название которой встречается чаще остальных, они проводят в интернете в среднем больше 120 минут и их возраст меньше 30 лет. 
Cохраните получившиеся строки в новый датафрейм `filtered_df`.


In [ ]:
# Ваш код здесь

***Задача 19***

Для получившегося датафрейма `filtered_df` выведите строки с данными пользователей, для которых либо не указано среднее время, проведённое на сайте, либо указанное время больше 55 минут. Cохраните получившиеся строки в новый датафрейм `filtered_df2`.


In [ ]:
# Ваш код здесь

**Дополнительные задачи**

***Задача 20***

Для датафрейма `adv123_df` выведите строки с данными, заголовок рекламного сообщения в которых включает слово `model`.

Для этого вам понадобится метод [`contains`](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.contains.html).

Cохраните получившиеся строки в новый датафрейм `model_df`.


In [ ]:
# Ваш код здесь

***Задача 21***

Для получившегося датафрейма `model_df` выведите строки с данными пользователей, которые проводят на сайте больше среднего времени по всем пользователям в данном датафрейме. Cохраните получившиеся строки в новый датафрейм `model_popular_df`.


In [ ]:
# Ваш код здесь

***Задача 22***

Для объединённого датафрейма `model_popular_df` выведите строки с данными пользователей, номера которых равны 501, 712 и 823. 

Для этого вам понадобится метод [`isin`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.isin.html).

*Подсказка:* Не забудьте, что номера пользователей находятся не в обычной колонке, а в колонке, являющейся индексом, поэтому вам необходимо получить значения индекса, а не обычной колонки.

In [ ]:
# Ваш код здесь